### DISCUSSION TITLE
[serializer] How to properly serialize an array of objects

### DISCUSSION THREAD

**Discussion by schlangz** [activej/activej#173] [Q&A]
Hi, I'm trying to serialize an array of objects (MyClass) but I'm only serializing an empty array without the objects.
Since deserialization works and there are no errors I'm not sure what I'm missing. Please advise, thank you.

Serializer:
```
    SerializerDefClass def = SerializerDefClass.of( MyClass.class );
    SerializerForType type = new SerializerDefBuilder.SerializerForType( MyClass.class, def );
    this.serializer = SerializerBuilder.create( this.classLoader )
                                       .withSerializer( MyClass.class, def )
                                       .build( MyClass[].class, new SerializerForType[] { type } );
```

Serialization:
```
    try (ByteArrayOutputStream bos = new ByteArrayOutputStream( BUFFER_SIZE );
                    StreamOutput output = StreamOutput.create( bos ))
    {
      this.serializer.encode( output.getBinaryOutput(), MyClassArray );
      return bos.toByteArray();
    }
```

Deserialization:
```
    try (ByteArrayInputStream bis = new ByteArrayInputStream( byte[] );
                    StreamInput input = StreamInput.create( bis ))
    {
      return this.serializer.decode( input.getBinaryInput() );
    }
```

**Comment by schlangz** [comment]
This is the generated serialier:
```
    public final int encode(final byte[] array, int writeVarInt, final Object o) {
        writeVarInt = BinaryOutputUtils.writeVarInt(array, writeVarInt, ((MyClass[])o).length);
        for (int length = ((MyClass[])o).length, i = 0; i < length; ++i) {}
        return writeVarInt;
    }
    
    public final void encode(final BinaryOutput binaryOutput, final Object o) {
        final int writeVarInt = BinaryOutputUtils.writeVarInt(binaryOutput.array(), binaryOutput.pos(), ((MyClass[])o).length);
        for (int length = ((MyClass[])o).length, i = 0; i < length; ++i) {}
        binaryOutput.pos(writeVarInt);
    }
    
    public final Object decode(final BinaryInput binaryInput) {
        final int varInt = binaryInput.readVarInt();
        final MyClass[] array = new MyClass[varInt];
        for (int n = varInt, i = 0; i < n; ++i) {
            array[i] = new MyClass();
        }
        return array;
    }
    
    public final Object decode(final byte[] array, final int n) {
        final int varInt = new BinaryInput(array, n).readVarInt();
        final MyClass[] array2 = new MyClass[varInt];
        for (int n2 = varInt, i = 0; i < n2; ++i) {
            array2[i] = new MyClass();
        }
        return array2;
    }
```

**Comment by schlangz** [comment]
If I manually execute these calls before calling builder.build(), everything works as expected:
```
    builder.scanSerializers( MyClass.class, defTypes );
    builder.scanAnnotations( MyClass.class, defTypes, def );
```

**Comment by eduard-vasinskyi** [comment]
Hello, @schlangz
First of all, to serialize data using [StreamInput](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/stream/StreamInput.java), you should call `StreamInput#serialize`, rather than writing directly to a [BinaryOutput](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/BinaryOutput.java). Also, you need to call `StreamInput#flush` before converting `ByteArrayOutputStream` to a byte array.
Similarly, you need to use `StreamOutput#deserialize` rather than reading from a [BinaryInput](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/BinaryInput.java)
 
As for serializing an array of objects, we suggest you use a container class to hold an array. Here is [a gist that demonstrates how to do just that.](https://gist.github.com/eduard-vasinskyi/6f5293428fc514ca23baaeecfa615792)

**Comment by schlangz** [comment]
Thank you for the response and the example, let me work with that.
